controllers/workspace/finalize.go

```diff
@@ -0,0 +1,213 @@
+}
+
+func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *v1alpha2.DevWorkspace) error {
+	if !hasFinalizer(workspace) {
+		return nil
+	}
+	// Need to make sure Deployment is cleaned up before starting job to avoid mounting issues for RWO PVCs
+	wait, err := provision.DeleteWorkspaceDeployment(ctx, workspace, r.Client)
+	if err != nil {
+		return err
+	}
+	if wait {
+		return nil
+	}
+	specJob, err := r.getSpecCleanupJob(workspace)
+	if err != nil {
+		return err
+	}
+	clusterJob, err := r.getClusterCleanupJob(ctx, workspace)
+	if err != nil {
+		return err
+	}
+	if clusterJob == nil {
+		err := r.Create(ctx, specJob)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+	if !equality.Semantic.DeepDerivative(specJob.Spec, clusterJob.Spec) {
+		err := r.Patch(ctx, specJob, client.Merge)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	for _, condition := range clusterJob.Status.Conditions {
+		if condition.Status != corev1.ConditionTrue {
+			continue
+		}
+		switch condition.Type {
+		case batchv1.JobComplete:
+			log.Info("PVC clean up job successful; clearing finalizer")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		case batchv1.JobFailed:
+			log.Error(fmt.Errorf("PVC clean up job failed: message %q", condition.Message),
+				"Failed to clean PVC on workspace deletion")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		}
+	}
+	return nil
+}
+
+func (r *DevWorkspaceReconciler) getSpecCleanupJob(workspace *v1alpha2.DevWorkspace) (*batchv1.Job, error) {
+	workspaceId := workspace.Status.WorkspaceId
+	pvcName := config.ControllerCfg.GetWorkspacePVCName()
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      common.PVCCleanupJobName(workspaceId),
+			Namespace: workspace.Namespace,
+			Labels: map[string]string{
+				config.WorkspaceIDLabel: workspaceId,
+			},
+		},
+		Spec: batchv1.JobSpec{
+			Completions:  &cleanupJobCompletions,
+			BackoffLimit: &cleanupJobBackoffLimit,
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					RestartPolicy:   "Never",
+					SecurityContext: provision.GetDevWorkspaceSecurityContext(),
+					Volumes: []corev1.Volume{
+						{
+							Name: pvcName,
+							VolumeSource: corev1.VolumeSource{
+								PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+									ClaimName: pvcName,
+								},
+							},
+						},
+					},
+					Containers: []corev1.Container{
+						{
+							Name:    common.PVCCleanupJobName(workspaceId),
+							Image:   images.GetPVCCleanupJobImage(),
+							Command: []string{"/bin/sh"},
+							Args: []string{
+								"-c",
+								fmt.Sprintf(cleanupCommandFmt, path.Join(pvcClaimMountPath, workspaceId)),
+							},
+							Resources: corev1.ResourceRequirements{
+								Limits: corev1.ResourceList{
+									corev1.ResourceMemory: resource.MustParse(config.PVCCleanupPodMemoryLimit),
```

Comment:
Is it really a good idea to panic here if mistake took a place in source code?

Maybe better to initialize this var in init func, or just use var instead of const and then

```
// PVCCleanupPodMemoryLimit is the memory limit used for PVC clean up pods
var PVCCleanupPodMemoryLimit = resource.MustParse("32Mi")
```

Reply:
Makes sense, thanks.